kyo-core/shared/src/main/scala/kyo/Fiber.scala

```diff
@@ -1,746 +1,527 @@
+
+    def fromFuture[A](future: => Future[A])(using Frame): Fiber[A, Any] < Sync =
         Sync.Unsafe(Unsafe.fromFuture(future))
 
-    /** Runs an asynchronous computation in a new Fiber.
-      *
-      * @param v
-      *   The computation to run
-      * @return
-      *   A Fiber representing the running computation
-      */
-    def init[E, A, S](
-        using isolate: Isolate.Contextual[S, Sync]
-    )(v: => A < (Abort[E] & Async & S))(using Frame): Fiber[E, A] < (Sync & S) =
+    def init[E, A, S, S2](
+        using isolate: Isolate[S, Sync, S2]
+    )(v: => A < (Abort[E] & Async & S))(using reduce: Reducible[Abort[E]], frame: Frame): Fiber[A, reduce.SReduced & S2] < (Sync & S) =
         isolate.runInternal((trace, context) =>
-            Fiber.fromTask(IOTask(v, trace, context))
+            isolate.capture { state =>
+                val io = isolate.isolate(state, v).map(r => Kyo.lift(isolate.restore(r)))
+                IOTask(io, trace, context).asInstanceOf[Fiber[A, reduce.SReduced & S2]]
+            }
         )
 
-    private def result[E, A](result: Result[E, A]): Fiber[E, A] = IOPromise(result)
-
-    extension [E, A](self: Fiber[E, A])
-
-        private[kyo] def asPromise: IOPromise[E, A] = self.asInstanceOf[IOPromise[E, A]]
+    extension [A, S](self: Fiber[A, S])
+        def done(using Frame): Boolean < Sync =
+            Sync.Unsafe(Unsafe.done(self)())
+        def map[B](f: A => B < Sync)(using Frame): Fiber[B, S] < Sync =
+            Sync.Unsafe(Unsafe.map(self)((r => Sync.Unsafe.evalOrThrow(f(r)))))
+        def flatMap[B, S2](f: A < S => Fiber[B, S2] < Sync)(using Frame): Fiber[B, S & S2] < Sync =
+            Sync.Unsafe(Unsafe.flatMap(self)(r => Sync.Unsafe.evalOrThrow(f(r))))
+        def mask(using Frame): Fiber[A, S] < Sync =
+            Sync.Unsafe(Unsafe.mask(self)())
+        def waiters(using Frame): Int < Sync =
+            Sync.Unsafe(Unsafe.waiters(self)())
+        def unsafe: Fiber.Unsafe[A, S] =
+            self
+    end extension
 
-        /** Gets the result of the Fiber.
-          *
-          * @return
-          *   The result of the Fiber
-          */
-        def get(using reduce: Reducible[Abort[E]], frame: Frame): A < (reduce.SReduced & Async) =
-            Async.get(self.asPromise)
+    extension [E, A, S](self: Fiber[A, Abort[E] & S])
+        private[kyo] def lower: IOPromise[E, A < S] = self.asInstanceOf[IOPromise[E, A < S]]
 
-        /** Uses the result of the Fiber to compute a new value.
-          *
-          * @param f
-          *   The function to apply to the Fiber's result
-          * @return
-          *   The result of applying the function to the Fiber's result
-          */
-        def use[B, S](f: A => B < S)(using reduce: Reducible[Abort[E]], frame: Frame): B < (reduce.SReduced & Async & S) =
-            Async.use(self.asPromise)(f)
+        def get(using reduce: Reducible[Abort[E]], frame: Frame): A < (reduce.SReduced & Async & S) =
```

Comment:
is Reducible usefull here?

Reply:
good point! it isn't since `Async` includes `Abort[Nothing]`. Removed